ethcore/engine/src/engine.rs

```diff
@@ -331,6 +331,12 @@ pub trait Engine: Sync + Send {
 		Ok(*header.author())
 	}
 
+	/// Overrides the block gas limit. Whenever this returns `Some` for a header, the next block's gas limit must be
+	/// exactly that value.
+	fn gas_limit_override(&self, _header: &Header) -> Option<U256> {
```

Comment:
So I think I made this suggestion before but perhaps it got lost.

In order to keep the `Engine` trait from growing ever fatter, how about we have a `fn gas_limits(&self) -> Option<(U256, U256)>` that return the min/max limit. The default impl could return  `(self.params().min_gas_limit, self.params().max_gas_limit)` and then Aura can provide its own impl with the limit override logic tucked away nicely. This would require refactoring other code around the project but lets us remove `maximum_gas_limit(&self)` and saves us from adding two new methods.

Not sure if you tried this and it doesn't work or if you feel it's a bad idea?

Reply:
Sorry, I missed that!

I'm not sure about this idea: When the engine changes the gas limit, we don't want to do [these checks](https://github.com/paritytech/parity-ethereum/pull/10928/files#diff-4eb43f14344f39e06bd7ceb98f5b3353R345):

At the moment, it seems to be impossible to produce a valid block if the interval based on the gas limit divisor doesn't overlap with the interval based on the engine's min and max gas limits.
Maybe engines that use a gas limit contract should just use a gas limit divisor of 1, but that would still prevent them from more than doubling the gas limit from one block to the next.

But we could make `gas_limits` return a `bool` as well, that tells us whether it's an override. Or it could return an enum?
```rust
enum GasLimits {
    MinMax(U256, U256),
    Override(U256),
}
```

Reply:
Hmm, so why do you need to know that there was an override? Why isn't it enough to know what the bounds must be at that block? In this PR there are 3 error cases: gas too low, too high, overridden but mismatching – why is it important to know that the error came from an override?

Reply:
There seem to be two different intervals for the block gas limit:

* [the engine's min/max](https://github.com/paritytech/parity-ethereum/pull/10928/files#diff-4eb43f14344f39e06bd7ceb98f5b3353R331) and
* [the range calculated using the gas limit divisor](https://github.com/paritytech/parity-ethereum/pull/10928/files#diff-4eb43f14344f39e06bd7ceb98f5b3353R347), which allows us to change the limit only by a certain percentage compared to _the previous block_.

The second limitation can conflict with the first, if the engine makes sudden changes to the gas limit. So I think it should be disabled if there is an override.

I'm not really sure at all what's the best way to approach this. To give you some context, the gist is:
In [POSDAO](https://forum.poa.network/c/posdao), there will be rather expensive implicit calls to the governance contracts once per week, to distribute rewards and select new validators. While these are ongoing, we want to limit the gas available for transactions, so that the total CPU usage per block remains roughly constant. (For details see: https://github.com/poanetwork/parity-ethereum/issues/119#issue-430516284)